### PR TITLE
JAVA-2717: Add OSS Cassandra 4.0 to build.yaml (3.x)

### PR DESCRIPTION
**Pull request by emerkle826**
This PR simply adds Cassandra 4.0 to the build matrix on the old Jenkins server that is triggered via Job Creator:
https://jenkins-drivers-old.build.dsinternal.org/view/java-driver/job/datastax.java-driver.java2717.commit/

We'll need either this PR merged <b>OR</b> merge the [PR here](https://github.com/datastax/java-driver/pull/1420) if we want to switch the 3.x branch over to CloudBees

### CODE REVIEW

File: driver-core/src/test/java/com/datastax/driver/core/CCMBridge.java [hunk 1]
```diff
@@ -1089,7 +1089,8 @@ public void run() {
     }
 
     private static boolean isThriftSupported(VersionNumber cassandraVersion) {
-      return cassandraVersion.compareTo(VersionNumber.parse("4.0")) < 0;
+      // Thrift is removed from some pre-release 4.x versions, make the comparison work for those
+      return cassandraVersion.compareTo(VersionNumber.parse("4.0-a")) < 0;
```
Comment (adutra) [inline driver-core/src/test/java/com/datastax/driver/core/CCMBridge.java]: You can use `nextStable()` for that:

```
 return cassandraVersion.nextStable().compareTo(VersionNumber.parse("4.0")) < 0;
```

File: driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java [hunk 2]
```diff
@@ -258,14 +258,19 @@ protected StringBuilder appendOptions(StringBuilder sb, boolean formatted) {
...
-    and(sb, formatted)
-        .append("dclocal_read_repair_chance = ")
-        .append(options.getLocalReadRepairChance());
+    if (cassandraVersion.getMajor() < 4)
+      sb.append("read_repair_chance = ").append(options.getReadRepairChance());
+    else sb.append("read_repair = 'BLOCKING'");
+    if (cassandraVersion.getMajor() < 4)
+      and(sb, formatted)
+          .append("dclocal_read_repair_chance = ")
+          .append(options.getLocalReadRepairChance());
```
Comment (emerkle826) [inline driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java:267]: This was done to get tests passing against C* 4.0. I don't think this is the right way to address this as the changed table options are just hard-coded here, with no way to change them or retrieve them in a TableMetadataOptions object.

File: driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java [hunk 3]
```diff
@@ -258,14 +258,19 @@ protected StringBuilder appendOptions(StringBuilder sb, boolean formatted) {
...
     if (cassandraVersion.getMajor() < 2
         || (cassandraVersion.getMajor() == 2 && cassandraVersion.getMinor() == 0))
       and(sb, formatted).append("replicate_on_write = ").append(options.getReplicateOnWrite());
     and(sb, formatted).append("gc_grace_seconds = ").append(options.getGcGraceInSeconds());
+    if (cassandraVersion.getMajor() > 3)
+      and(sb, formatted).append("additional_write_policy = '99p'");
```
Comment (emerkle826) [inline driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java:273]: This is another case like the above. Just a hard-coded value.

File: driver-core/src/test/java/com/datastax/driver/core/TestUtils.java [hunk 4]
```diff
@@ -986,7 +986,7 @@ public static Level setLogLevel(String logger, Level newLevel) {
    * @param ccm cluster to check against
    */
   public static void compactStorageSupportCheck(CCMAccess ccm) {
-    if (ccm.getCassandraVersion().compareTo(VersionNumber.parse("4.0.0")) >= 0) {
+    if (ccm.getCassandraVersion().compareTo(VersionNumber.parse("4.0-a")) >= 0) {
```
Comment (emerkle826) [inline driver-core/src/test/java/com/datastax/driver/core/TestUtils.java]: this needs to be changed to 
```java
if (ccm.getCassandraVersion().nextStabel().compareTo(VersionNumber.parse("4.0")) >= 0) {
```

File: driver-tests/osgi/src/test/java/com/datastax/driver/osgi/BundleOptions.java [hunk 5]
```diff
@@ -165,7 +165,17 @@ public static CompositeOption defaultOptions() {
                 mavenBundle("ch.qos.logback", "logback-classic", getVersion("logback.version")),
                 mavenBundle("ch.qos.logback", "logback-core", getVersion("logback.version")),
                 mavenBundle("io.dropwizard.metrics", "metrics-core", getVersion("metrics.version")),
-                mavenBundle("org.yaml", "snakeyaml", getVersion("snakeyaml.version")),
+                // mavenBundle("org.yaml", "snakeyaml", getVersion("snakeyaml.version")),
```
Comment (emerkle826) [inline driver-tests/osgi/src/test/java/com/datastax/driver/osgi/BundleOptions.java]: need to removed the commented line

File: driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java (hunk 2 above)
Comment (emerkle826) [reply driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java:267]: created https://datastax-oss.atlassian.net/browse/JAVA-2730 for this

Comment (olim7t) [reply driver-core/src/main/java/com/datastax/driver/core/AbstractTableMetadata.java:267]: We also need to amend the code that parses the options. That's done in the constructor of `TableOptionsMetadata`.

As far as outputting the options conditionally, I think that's the right way to do it. That's how it was already handled previously (driver 3 hard-codes a lot more things).

**Comment by emerkle826** [comment]
Closing this in favor of https://github.com/datastax/java-driver/pull/1420

**Pull request closed without merging**